src/creature_instances.c

```diff
@@ -1053,4 +1155,243 @@ void delay_heal_sleep(struct Thing *creatng)
+    int creature_idx = dungeon->creatr_list_start;
+    int k = 0;
+    while (creature_idx != 0 && (*found_count) < MAX_CREATURES_SEARCHED)
+    {
+        struct Thing* candidate = thing_get(creature_idx);
+        struct CreatureControl* cctrl = creature_control_get_from_thing(candidate);
+        if (creature_control_invalid(cctrl))
+        {
+            ERRORLOG("Invalid creature control");
+            ok = false;
+            break;
+        }
+
+        creature_idx = cctrl->players_next_creature_idx;
+
+        const struct InstanceInfo* inst_inf = creature_instance_info_get(inst_idx);
+        if (inst_inf->validate_func_idx[1] > 0)
+        {
+            if(!creature_instances_validate_func_list[inst_inf->validate_func_idx[1]](source, candidate, inst_idx))
+            {
+                // This candidate is out.
+                continue;
+            }
+        }
+
+        // @todo Consider checking thing_in_field_of_view() in the future, now it is buggy.
+        // We assume that the source must see the target before it can cast the spell.
+        int range = get_combat_distance(source, candidate);
+        if (range < inst_inf->range_min || range > inst_inf->range_max ||
+            !creature_can_see_combat_path(source, candidate, range))
+        {
+            continue;
+        }
+
+        targets[*found_count] = candidate;
+        (*found_count)++;
+
+        k++;
+        if (k > CREATURES_COUNT)
+        {
+            ERRORLOG("Infinite loop detected when sweeping creatures list");
+            ok = false;
+            break;
+        }
+    }
+
+    return ok;
+}
+
+/**
+ * @brief Search the suitable targets for CrInst_RANGED_HEAL
+ *
+ * @param source The source creature
+ * @param inst_idx The spell instance index
+ * @param targets The list of the found creatures. Caller must free this.
+ * @param found_count The number of the found creatures.
+ * @return TbBool True if there is no error, false if otherwise
+ */
+TbBool search_target_ranged_heal(struct Thing *source, CrInstance inst_idx, struct Thing **targets, short *found_count)
+{
+    if (!targets || !found_count)
+    {
+        ERRORLOG("Invalid parameters!");
+        return false;
+    }
+
+    if (!search_target_generic(source, inst_idx, targets, found_count))
+    {
+        return false;
+    }
+
+    struct Thing *best_choice = NULL;
+    TbBool ok = true;
+    int i = 0;
+    for (; i < *found_count; i++)
+    {
+        struct Thing *candidate = targets[i];
+        if (thing_is_invalid(candidate))
+        {
+            continue;
+        }
+        struct CreatureControl* cctrl = creature_control_get_from_thing(candidate);
+        if (creature_control_invalid(cctrl))
+        {
+            continue;
+        }
+        // Note that we only allow one target. No group buff are allowed yet.
+        // So, we only pick the weakest one creature here.
+        float hp_p_best = 0;
```

Comment:
Please try to avoid floats. You can use regular variables and simply multiply with 256 to get a ratio.

Reply:
OK.